runtime/compiler/runtime/JITServerAOTCache.cpp

```diff
@@ -745,6 +779,93 @@ JITServerAOTCache::printStats(FILE *f) const
    );
    }
 
+// Write the at most numRecordsToWrite to the given stream from the linked list starting at head.
+static bool
+writeAOTRecordList(FILE *f, AOTCacheRecord *head, size_t numRecordsToWrite)
+   {
+   AOTCacheRecord *w = head;
+   size_t recordsWritten = 0;
+   while (w && (recordsWritten < numRecordsToWrite))
+      {
+      const AOTSerializationRecord *record = w->dataAddr();
+      if (1 != fwrite(record, record->size(), 1, f))
+         {
+         return false;
+         }
+      ++recordsWritten;
+      w = w->getNextRecord();
+      }
+
+   return true;
+   }
+
+static bool
+writeCachedMethodList(FILE *f, CachedAOTMethod *head, size_t numRecordsToWrite)
+   {
+   CachedAOTMethod *w = head;
+   size_t recordsWritten = 0;
+   while (w && (recordsWritten < numRecordsToWrite))
+      {
+      SerializedAOTMethod *record = &w->data();
+      if (1 != fwrite(record, record->size(), 1, f))
+         {
+         return false;
+         }
+      ++recordsWritten;
+      w = w->getNextRecord();
+      }
+
+   return true;
+   }
+
+static void getCurrentAOTCacheVersion(JITServerAOTCacheVersion *version)
+   {
+   memcpy(version->eyeCatcher, JITSERVER_AOTCACHE_EYECATCHER, JITSERVER_AOTCACHE_EYECATCHER_LENGTH);
+   version->snapshotVersion = JITSERVER_AOTCACHE_VERSION;
+   version->jitserverVersion = JITServer::CommunicationStream::getJITServerVersion();
+   version->serverUID = TR::CompilationInfo::get()->getPersistentInfo()->getServerUID();
+   }
+
+// Write a full AOT cache snapshot to a stream. After the header information, the
+// AOTSerializationRecord or SerializedAOTMethod data (depending on record type) in each
+// record traversal is written directly to the stream in sections, since the full AOT record
+// can be reconstructed from only this information. These sections are ordered so that, when
+// reading the snapshot, the dependencies of each record will already have been read by the
+// time we get to that record.
+bool
+JITServerAOTCache::writeCache(FILE *f) const
+   {
+   JITServerAOTCacheHeader header;
+   getCurrentAOTCacheVersion(&header.version);
+   header.numClassLoaderRecords = _classLoaderMap.size();
```

Comment:
To be precise, you need to acquire the mutex before reading `size()`.
Also, I am thinking it may be better to read the sizes in reverse order. In the order here it's possible to read some number of class loader records, then the thread get scheduled out, some other threads add some CachedAOTMethods increasing their size, and our thread will write to disk these methhods that may not have dependent records (class loader records and such).

Reply:
True, the sizes must be read in reverse dependency order, I missed that. While acquiring the locks is not strictly necessary per se to read the sizes (stored in a single `size_t` field in any reasonable implementation), it has the benefit of the implicit read barriers that will guarantee the order of reads. So each `size()` value should be read inside a corresponding critical section.